lib/mapbox/matrix.rb

```diff
@@ -0,0 +1,47 @@
+module Mapbox
+  class Matrix
+    include Mapbox::APIOperations::Request
+    extend Mapbox::HashUtils
+
+    def self.assemble_params(options={})
+      annotations = options[:annotations]
+      approaches = options[:approaches]
+      destinations = options[:destinations]
```

Comment:
@Bathlamos Not sure I totally follow - apologies. Do you mind clarifying what you mean here?

Reply:
For instance, for the parameter `destinations`, the documentation says `Use the coordinates at a given index as destinations. Possible values are: a semicolon-separated list of 0-based indices, or all (default).`

So technically, both of the following could be true:
`destinations=[0,1]`
`destinations="all"`

However, your code doesn't handle the second case. Does that help?

Reply:
@Bathlamos Thanks! What do you think about this approach? https://github.com/mapbox/mapbox-sdk-rb/pull/39/commits/6b6e1f2c58f9d9991b79078e8dfafca71792824e

If you think it's OK, I can add it to other params that this is a problem for. Otherwise, would love to hear if you know of a better approach.

Reply:
@samfader This works, but keep in mind that it can fail in a subtle way. Let me give you an example:

If you set `destinations = "all"`, then your if statement at https://github.com/mapbox/mapbox-sdk-rb/blob/6b6e1f2c58f9d9991b79078e8dfafca71792824e/lib/mapbox/matrix.rb#L9 doesn't get executed, which is expected. What is not however is that `destinations` gets excluded from `opts` at https://github.com/mapbox/mapbox-sdk-rb/blob/6b6e1f2c58f9d9991b79078e8dfafca71792824e/lib/mapbox/matrix.rb#L15 because you still filter on the key.

If you want, you could just use an approach similar to this, which could be a bit more compact. FYI I haven't tested this code:

```
module Mapbox
  class Matrix
    include Mapbox::APIOperations::Request
    extend Mapbox::HashUtils

    def self.assemble_params(options={})
      opts = dup(options)

      self.joinArrayParam(opts, :destinations)
      self.joinArrayParam(opts, :annotations)
      self.joinArrayParam(opts, :approaches)
      self.joinArrayParam(opts, :sources)

      return "?#{URI.encode_www_form(opts)}"
    end

    # Share in a common file somewhere
    def self.joinArrayParam(opts, name, joinSymbol = ';')
      if opts[name].kind_of?(Array)
        opts[name] = opts[name].join(joinSymbol)
      end
    end

    def self.matrix(waypoints, profile, options={})
      formatted_waypoints = waypoints.map {|p| xy_from_hash(p).join ','}.join ';'
      params = self.assemble_params(options)
      
      return request(
          :get,
          "/directions-matrix/v1/mapbox/#{profile}/#{formatted_waypoints}.json#{params}",
          nil)
    end
  end
end
```

Reply:
Thanks @Bathlamos. I had to change your sample around a bit since **annotations** requires commas, not semicolons. How does this look? All tests passing locally.

https://github.com/mapbox/mapbox-sdk-rb/pull/39/commits/4d5d1f944ad96f4c7433ff1457a1b029b3f0f445

Reply:
This is interesting because now the method `joinArrayParam` needs to be aware of the context. If you pull the logic from the joinSymbol out, like as follows, it'd be a bit easier. Just a thought!

```ruby
self.joinArrayParam(opts, :destinations)
# here you can simply tell the method about your different join character
self.joinArrayParam(opts, :annotations, ',')
#self.joinArrayParam(opts, :approaches)
self.joinArrayParam(opts, :sources)

def self.joinArrayParam(opts, name, joinSymbol = ';')
  # This method shouldn't have behavior that's specific to the name parameter, for simplicity.
  if opts[name].kind_of?(Array)
    opts[name] = opts[name].join(joinSymbol)
  end
end
```

Also, it's nice to duplicate `opts` at the beginning of the method, since modifying parameters passed by reference isn't a best practice.

Reply:
@Bathlamos Thanks so much for your patience/helping me out here. How does the latest look? I changed around the way I was calling dup, [per the docs](https://ruby-doc.org/core-2.6.1/Object.html#method-i-dup), since I was getting an error. Otherwise I think this looks a lot better.